### PR TITLE
planner: move index advisor into the kernel

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #12303

Problem Summary: planner: move index advisor into the kernel

### What changed and how does it work?

planner: move index advisor into the kernel

1. move the algorithm code into TiDB.
2. add more test cases

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/indexadvisor/algorithm.go [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+	s "github.com/pingcap/tidb/pkg/util/set"
+)
+
+/*
+	This algorithm resembles the index selection algorithm published in 1997 by Chaudhuri
+	and Narasayya. Details can be found in the original paper:
+	Surajit Chaudhuri, Vivek R. Narasayya: An Efficient Cost-Driven Index Selection
+	Tool for Microsoft Query Server. VLDB 1997: 146-155
+	This implementation is the Golang version of
+	github.com/hyrise/index_selection_evaluation/blob/refactoring/selection/algorithms/auto_admin_algorithm.py.
```
Comment (hawkingrei) [inline pkg/planner/indexadvisor/algorithm.go:31]: The IDE can treat this as a URL, allowing it to be clicked and opened directly.

```suggestion
	https://github.com/hyrise/index_selection_evaluation/blob/refactoring/selection/algorithms/auto_admin_algorithm.py.
```

File: pkg/planner/indexadvisor/indexadvisor.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	defer func() {
+		if r := recover(); r != nil {
+			l().Error("panic in AdviseIndexes", zap.Any("recover", r))
+			err = fmt.Errorf("panic in AdviseIndexes: %v", r)
+		}
+	}()
+
+	// prepare what-if optimizer
+	opt := NewOptimizer(sctx)
+	l().Info("what-if optimizer prepared")
```
Comment (hawkingrei) [inline pkg/planner/indexadvisor/indexadvisor.go:59]: ```l``` is too hard to understand this meaning. We can call it ```indexAdvisorLog```

**Review by hawkingrei — COMMENTED** [review]
Other, LGTM

File: pkg/planner/indexadvisor/indexadvisor.go (hunk 2 above)
Comment (qw4990) [reply pkg/planner/indexadvisor/indexadvisor.go]: Thanks for this suggestion, I've updated it.

**Review by elsa0520 — APPROVED** [review]
LGTM

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**